src/module.c

```diff
@@ -6038,6 +6023,103 @@ RedisModuleCallReply *RM_Call(RedisModuleCtx *ctx, const char *cmdname, const ch
+ *     * `C` -- Check if command can be executed according to ACL rules.
+ *     * `S` -- Run the command in a script mode, this means that it will raise
+ *              an error if a command which are not allowed inside a script
+ *              (flagged with the `deny-script` flag) is invoked (like SHUTDOWN).
+ *              In addition, on script mode, write commands are not allowed if there are
+ *              not enough good replicas (as configured with `min-replicas-to-write`)
+ *              or when the server is unable to persist to the disk.
+ *     * `W` -- Do not allow to run any write command (flagged with the `write` flag).
+ *     * `M` -- Do not allow `deny-oom` flagged commands when over the memory limit.
+ *     * `E` -- Return error as RedisModuleCallReply. If there is an error before
+ *              invoking the command, the error is returned using errno mechanism.
+ *              This flag allows to get the error also as an error CallReply with
+ *              relevant error message.
+ * * **...**: The actual arguments to the Redis command.
+ *
+ * On success a RedisModuleCallReply object is returned, otherwise
+ * NULL is returned and errno is set to the following values:
+ *
+ * * EBADF: wrong format specifier.
+ * * EINVAL: wrong command arity.
+ * * ENOENT: command does not exist.
+ * * EPERM: operation in Cluster instance with key in non local slot.
+ * * EROFS: operation in Cluster instance when a write command is sent
+ *          in a readonly state.
+ * * ENETDOWN: operation in Cluster instance when cluster is down.
+ * * ENOTSUP: No ACL user for the specified module context
+ * * EACCES: Command cannot be executed, according to ACL rules
+ * * ENOSPC: Write or deny-oom command is not allowed
+ * * ESPIPE: Command not allowed on script mode
+ *
+ * Example code fragment:
+ *
+ *      reply = RedisModule_CallWithUser(ctx, rm_user, "INCRBY","sc",argv[1],"10");
+ *      if (RedisModule_CallReplyType(reply) == REDISMODULE_REPLY_INTEGER) {
+ *        long long myval = RedisModule_CallReplyInteger(reply);
+ *        // Do something with myval.
+ *      }
+ *
+ * This API is documented here: https://redis.io/topics/modules-intro
+ */
+RedisModuleCallReply *RM_Call(RedisModuleCtx *ctx, const char *cmdname, const char *fmt, ...) {
+    robj **argv = NULL;
+    va_list ap;
+    int argc = 0, argv_len = 0, flags = 0;
+
+    /* Handle arguments. */
+    va_start(ap, fmt);
+    argv = moduleCreateArgvFromUserFormat(cmdname,fmt,&argc,&argv_len,&flags, ap);
+    va_end(ap);
+
+    return CallWithUserInternal(ctx, NULL, argv, argc, argv_len, flags);
+}
+
+/* same as RM_Call, but executed as the provided user with ACL enforcement per that user
+ * If called with a NULL user, returns an error and sets errno to EINVAL
+ */
+RedisModuleCallReply *RM_CallWithUser(RedisModuleCtx *ctx, RedisModuleUser *rm_user, const char *cmdname, const char *fmt, ...) {
```

Comment:
maybe it would be more flexible to expose an RM_SetCurrentUser (to mess with ctx->client, and then revert is when the context is destroyed), and then use a plain RM_Call?
if we expose this capability, there's no need for a new API, and it also enables other APIs that count on the current user to be affected by this.

Reply:
so, its possible, we do get ACLs out of that already.  In the future we might think we might need a greater ability, what I call CallWithClient.  i.e. the ability of modules to manage the lifcycle of clients themselves.  this would be for watch/multi/exec to work correctly for us to be able to have consistent "client session" mapping across all nodes in the cluster.

i.e. a client issues a watch, it needs to be marked dirty and have exec fail consistently across all nodes.  as redis already does that at the client object level, this seems to be the best way to do it, but that's future work.

Reply:
even then, we could maybe change the client that's used by the context too (same as we can change the user).
and there's no need for a specific RM_Call variant (and possibly other APIs) for these.

Reply:
so I sort of did this, no messing with ctx->client, messing with ctx and using that to pass to the 'temp client' that is allocated in rm_call

Reply:
not resolving this yet, as unsure if what I did was what you want